File: src/main/java/com/github/jacekpoz/common/sendables/database/queries/user/GetUserQuery.java
package com.github.jacekpoz.common.sendables.database.queries.user;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.github.jacekpoz.common.sendables.database.queries.basequeries.UserQuery;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@ToString(callSuper = true)
@EqualsAndHashCode(callSuper = true)
public class GetUserQuery extends UserQuery {

    @Getter
    private final String username;

    @JsonCreator
    private GetUserQuery(
            @JsonProperty("userID") long userID,
            @JsonProperty("username") String username,
            @JsonProperty("callerID") long callerID
    ) {
        super(userID, callerID);
        this.username = username;
    }

    public GetUserQuery(long userID, long callerID) {
        this(userID, null, callerID);
    }

    public GetUserQuery(String username, long callerID) {
        this(-1, username, callerID);
    }

}

File: src/main/java/com/github/jacekpoz/common/sendables/database/queries/user/GetUserQuery.kt
package com.github.jacekpoz.common.sendables.database.queries.user

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.jacekpoz.common.sendables.database.queries.basequeries.UserQuery

open class GetUserQuery @JsonCreator constructor(
    @JsonProperty("userID") userID: Long = -1,
    @JsonProperty("username") val username: String? = null,
    @JsonProperty("callerID") callerID: Long,
) : UserQuery(userID, callerID) {

    constructor(
        userID: Long,
        callerID: Long,
    ) : this(userID, null, callerID)

    constructor(
        username: String?,
        callerID: Long,
    ) : this(-1, username, callerID)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is GetUserQuery) return false
        if (!super.equals(other)) return false
        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + (username?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String = "GetUserQuery(userID=$userID, username='$username', callerID=$callerID)"

}